File: misc/AECTech2024/02_Command/PaintHeights.py
#! python3
# https://www.nuget.org/packages/EasingFunctions
# https://github.com/emmauss/easing-functions
#r "nuget: EasingFunctions, 1.0.1"
import sys
import System
from System.Drawing import Color

import Rhino
from Rhino.Input import RhinoGet
from Rhino.Input.Custom import GetPoint, OptionDouble
from Rhino.DocObjects import ObjectType
from Rhino.Commands import Result
from Rhino.Geometry import Point3d, Ray3d, Circle, Plane, Transform, Line

from Eto.Forms import Keyboard, Keys

from Easing import Quartic, Vector

class Brush(GetPoint):
    def __init__(self, brush_surface):
        super().__init__()
        self.SetCommandPrompt("Paint")

        self._bsurface = brush_surface

        # Default values
        self._size = 68.0
        self._falloff = 128.0
        self._strength = 0.5
        self._maxValue = 60.0
        self._minValue = 1.0
        self._mouseDown = False

        self._ease = Quartic(Vector(self._falloff - self._size, self._strength))

        # Options with range
        self.optsizedouble = OptionDouble(self._size, 1.0, 100.0)
        self.optfalloffdouble = OptionDouble(self._falloff, 1.0, 100.0)
        self.optstrengthdouble = OptionDouble(self._strength, 0.1, 10.0)

        # Adding options
        self.sizeIndex, _ = self.AddOptionDouble("BrushSize", self.optsizedouble)
        self.fosizeIndex, _ = self.AddOptionDouble("FalloffSize", self.optfalloffdouble)
        self.strengthIndex, _ = self.AddOptionDouble("Strength", self.optstrengthdouble)

    def DoPaint(self):
        res = None
        while res != Rhino.Input.GetResult.Cancel:
            res = self.Get(True)
            if res == Rhino.Input.GetResult.Option:
                opt = self.Option()
                if opt.Index == self.sizeIndex:
                    self._size = self.optsizedouble.CurrentValue
                elif opt.Index == self.fosizeIndex:
                    self._falloff = self.optfalloffdouble.CurrentValue
                elif opt.Index == self.strengthIndex:
                    self._strength = self.optstrengthdouble.CurrentValue
                self._ease = Quartic(Vector(self._falloff - self._size, self._strength));

    def OnMouseMove(self, e):
        self._mouseDown = e.LeftButtonDown

    def OnDynamicDraw(self, e):
        if not e.Viewport:
            return
        
        p = e.CurrentPoint
        cp = e.Viewport.WorldToClient(p)
        success, line = e.Viewport.GetFrustumLine(cp.X, cp.Y)
        if success:
            line.Flip()
            ray = Ray3d(line.From, line.Direction)

            closest_on_mesh = sys.maxsize
            head = None
            falloff = None

            for mesh in self._bsurface.GetMeshes(Rhino.Geometry.MeshType.Render):
                t = Rhino.Geometry.Intersect.Intersection.MeshRay(mesh, ray)
                if t >= 0:
                    closest_on_mesh = t
                    hit_point = ray.PointAt(t)
                    mp = mesh.ClosestMeshPoint(hit_point, sys.maxsize)
                    norm = mesh.NormalAt(mp)
                    head_plane = Plane(hit_point, norm)
                    head = Circle(head_plane, self._size)
                    falloff = Circle(head_plane, self._falloff)

            if closest_on_mesh != sys.maxsize:
                e.Display.DrawCircle(head, Color.Blue, 3)
                if self._falloff != self._size:
                    e.Display.DrawCircle(falloff, Color.LightGray, 2)

            if self._mouseDown:
                reverse = Keyboard.Modifiers.HasFlag(Keys.Shift)
                self.ApplyBrush(e.RhinoDoc, p, reverse)

    def ApplyBrush(self, doc, at, reverse):
        for robj in doc.Objects.FindByObjectType(ObjectType.Extrusion):
            bbox = robj.Geometry.GetBoundingBox(True)
            pp = bbox.Min
            dist = at.DistanceTo(pp)
            if dist < self._falloff:
                height = bbox.Max.Z - bbox.Min.Z
                value = self.ComputeValue(dist, height, reverse)
                xform = Transform.Scale(Plane.WorldXY, 1.0, 1.0, value / height)
                doc.Objects.Transform(robj, xform, True)

    def ComputeValue(self, dist, start, reverse):
        value = start
        incr = self._strength if dist < self._size else self._ease.Out(self._falloff - dist);
        value += -incr if reverse else incr
        value = min(max(value, self._minValue), self._maxValue)
        return value


success, obj_ref = RhinoGet.GetOneObject("Select Brush Plane", False, filter=ObjectType.Brep | ObjectType.Mesh)
if success == Result.Success:
    brush_surface = obj_ref.Object()
    brush = Brush(brush_surface)
    brush.DoPaint()

File: misc/AECTech2024/04_Publish/PaintHeights.cs
// #! csharp
// https://www.nuget.org/packages/EasingFunctions
// https://github.com/emmauss/easing-functions
#r "nuget: EasingFunctions, 1.0.1"
using System;
using System.Linq;
using SD = System.Drawing;
using EF = Eto.Forms;

using Rhino;
using Rhino.Input;
using Rhino.Input.Custom;
using Rhino.Geometry;
using Rhino.DocObjects;
using Rhino.Commands;

class Brush : GetPoint
{ 
    readonly RhinoObject _bsurface;

    // brush option indices
    readonly int sizeIndex;
    readonly int fosizeIndex;
    readonly int strengthIndex;
    
    // default values for brush options
    double _size = 8f;
    double _falloff = 16f;
    double _strength = 0.1f;

    // brush effect range
    double _maxValue = 60f;
    double _minValue = 1f;
    
    bool _mouseDown = false;
    Easing.Ease _ease;

    public Brush(RhinoObject brushSurface)
    {
        this.SetCommandPrompt("Paint");

        _bsurface = brushSurface;
        _ease = new Easing.Quartic(new Easing.Vector(Convert.ToSingle(_falloff), Convert.ToSingle(_strength)));

        // define the options using default values, and a valid range for each
        var optsizedouble = new OptionDouble(_size, 1, 100);
        var optfalloffdouble = new OptionDouble(_falloff, 1, 100);
        var optstrengthdouble = new OptionDouble(_strength, 0.1, 10);

        // add the options to this getter instance
        sizeIndex     = this.AddOptionDouble("BrushSize", ref optsizedouble, "Enter Brush Size");
        fosizeIndex   = this.AddOptionDouble("FalloffSize", ref optfalloffdouble, "Enter Falloff Size");
        strengthIndex = this.AddOptionDouble("Strength", ref optstrengthdouble, "Enter Brush Strength");
    }

    public void DoPaint()
    {
        GetResult res;
        do 
        {
            res = this.Get(onMouseUp: true);
            switch (res)
            {
                case GetResult.Option:
                	// one of brush options was changed
                    var opt = this.Option();
                    if (opt.Index == sizeIndex)
                    {
                        _size = opt.CurrentNumericValue;
                    }
                    else
                    if (opt.Index == fosizeIndex)
                    {
                        _falloff = opt.CurrentNumericValue;
                        _ease = new Easing.Quartic(new Easing.Vector(Convert.ToSingle(_falloff), Convert.ToSingle(_strength)));
                    }
                    else
                    if (opt.Index == strengthIndex)
                    {
                        _strength = opt.CurrentNumericValue;
                        _ease = new Easing.Quartic(new Easing.Vector(Convert.ToSingle(_falloff), Convert.ToSingle(_strength)));
                    }
                    break;

                case GetResult.Point:
                    // use click in viewport; lets continue the loop
                    break;
            }
        } while (res != GetResult.Cancel); // user hit ESC key
    }

    protected override void OnMouseMove(GetPointMouseEventArgs e)
    {
        _mouseDown = e.LeftButtonDown;
    }

    protected override void OnDynamicDraw(GetPointDrawEventArgs e)
    {
        if (e.Viewport is null)
        {
            return;
        }

        Point3d p = e.CurrentPoint;
        Point2d cp = e.Viewport.WorldToClient(p);
        if (e.Viewport.GetFrustumLine(cp.X, cp.Y, out Line line))
        {
            line.Flip();
            Ray3d ray = new Ray3d(line.PointAt(0), line.Direction);

            double closestOnMesh = double.MaxValue;
            Circle head = default;
            Circle falloff = default;

            foreach (Mesh mesh in _bsurface.GetMeshes(MeshType.Render)) 
            {
                double t = Rhino.Geometry.Intersect.Intersection.MeshRay(mesh, ray);
                if (t >= 0)
                {
                    closestOnMesh = t;
                    Point3d hitPoint = ray.PointAt(t);
                    MeshPoint mp = mesh.ClosestMeshPoint(hitPoint, double.MaxValue);
                    Vector3d norm = mesh.NormalAt(mp);
                    Plane headPlane = new Plane(hitPoint, norm);
                    head = new Circle(headPlane, hitPoint, _size);
                    falloff = new Circle(headPlane, hitPoint, _falloff);
                }
            }

            if (closestOnMesh != double.MaxValue)
            {
                e.Display.DrawCircle(head, SD.Color.Blue, 3);
                if (_falloff != _size)
                {
                    e.Display.DrawCircle(falloff, SD.Color.LightGray, 2);
                }
            }

            if (_mouseDown)
            {
                bool reverse = EF.Keyboard.Modifiers.HasFlag(EF.Keys.Shift);
                ApplyBrush(e.RhinoDoc, p, reverse);
            }
        }
    }

    void ApplyBrush(RhinoDoc doc, Point3d at, bool reverse)
    {
        foreach (RhinoObject robj in doc.Objects
                                        .FindByObjectType(ObjectType.Extrusion))
        {
            var bbox = robj.Geometry.GetBoundingBox(true);
            var pp = bbox.Min;
            double dist = at.DistanceTo(pp);
            if (dist < _falloff)
            {
                double height = (bbox.Max - bbox.Min).Z;
                double value = ComputeIncrement(dist, height, reverse);

                var xform = Transform.Scale(Plane.WorldXY, 1.0d, 1.0d, value / height);
                doc.Objects.Transform(robj, xform, true);
            }
        }
    }

    double ComputeIncrement(double dist, double start, bool reverse)
    {
        double value = start;
        double incr = _strength * (dist < _size ? 1.0d : _ease.Out(Convert.ToSingle(dist)));
        value += reverse ? -incr : incr;
        value = value > _maxValue ? _maxValue : value;
        value = value < _minValue ? _minValue : value;

        return value;
    }
}


var s = RhinoGet.GetOneObject("Select Brush Plane", false, ObjectType.Brep | ObjectType.Mesh, out ObjRef objRef);
if (s == Result.Success)
{
    RhinoObject brushPlane = objRef.Object();
    var brush = new Brush(brushPlane);
    brush.DoPaint();
}
